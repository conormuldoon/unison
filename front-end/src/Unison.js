

import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import 'react-day-picker/lib/style.css';
import { formatDate } from 'react-day-picker/moment';
import './App.css';
import ARLocationComponent from './ARLocationComponent';
import { FORMAT, SELF } from './Constant';
import createSelector from './selectorFactory';
import { today, tomorrow, expandLink, problemConnecting } from './Util';
import HttpStatus from 'http-status-codes';

import { createLocationFactory, createRemoveFactory } from './closureFactory';

/**
 * Application component for Unison. Once mounted, it connects to the back-end to receive a list of the locations being tracked.
 * 
 * @component
 * 
 */
function Unison(props) {


  const [fromDate, setFromDate] = useState(today());
  const [toDate, setToDate] = useState(tomorrow());
  const [option, setOption] = useState(null);
  const [marker, setMarker] = useState(null);
  const [curLoc, setCurLoc] = useState(null);
  const [curVar, setCurVar] = useState(null);
  const [varOpt, setVarOpt] = useState(null);
  const [locationMap, setLocationMap] = useState(null);
  const [collectionModel, setCollectionModel] = useState(null);

  const obtainData = () => {

    class ResponseError extends Error {
      constructor(message, status) {
        super(message);
        this.status = status;
      }
    }

    function checkResponse(response, endpoint) {
      if (!response.ok)
        throw new ResponseError('Bad response for ' + endpoint, response.status);
    }

    async function requestFeatureCollection(uri) {
      const response = await fetch(uri, {
        method: 'GET',
        headers: new Headers({
          'Accept': 'application/geo+json'
        })
      });


      checkResponse(response, uri);

      const fc = await response.json();

      const locationArray = fc.features;
      const n = locationArray.length;

      const newOption = [];
      const newMarker = [];

      for (let i = 0; i < n; i++) {
        newOption.push(locationArray[i].properties.name);
        let pos = [locationArray[i].geometry.coordinates[1], locationArray[i].geometry.coordinates[0]];

        const properties = locationArray[i].properties;
        newMarker.push({ name: properties.name, position: pos });

      }

      if (n > 0) {

        setOption(newOption);
        setMarker(newMarker);


      } else {
        setOption(null);
        setMarker(null);
      }


    }

    const halGetHeader = {
      method: 'GET',
      headers: new Headers({
        'Accept': 'application/hal+json'
      }),

    }

    function clearCurrent() {
      setCurLoc(null);
      setCurVar(null);
      setVarOpt(null);
      setLocationMap(null);

    }

    async function requestCollectionHAL(uri) {

      const response = await fetch(uri, halGetHeader);


      checkResponse(response, uri);

      const model = await response.json();


      setCollectionModel(model);

      if (model._embedded) {
        const list = model._embedded.locationModelList;
        const n = list.length;
        const map = new Map();
        for (let i = 0; i < n; i++) {
          map.set(list[i].name, list[i]);
        }

        setLocationMap(map);

        const varOpt = [];

        if (n > 0) {

          for (const rel in list[0]._links) {
            if (SELF === rel) {
              continue;
            }

            const s = rel.charAt(0).toUpperCase() + rel.substring(1);
            let optName = '';
            let sw = 0;
            for (let i = 1; i < s.length; i++) {
              const c = s.charAt(i);
              if (c >= 'A' && c <= 'Z') {
                optName += s.substring(sw, i) + ' ';
                sw = i;
              }
            }
            optName += s.substring(sw);
            varOpt.push(optName);
          }
        }
        if (varOpt.length > 0) {
          setCurLoc(list[0]);
          setCurVar(varOpt[0]);
          setVarOpt(varOpt);
          return;

        } else {

          clearCurrent();
          alert("No weather data is associated with the collection.")
        }


      } else {
        clearCurrent();
      }


    }

    async function requestModel() {
      try {

        const root = process.env.PUBLIC_URL;

        const response = await fetch(root, halGetHeader);

        checkResponse(response, root);


        const model = await response.json();

        const uri = model._links.locationCollection.href;
        await requestCollectionHAL(uri);
        await requestFeatureCollection(uri);

      } catch (e) {
        if (e instanceof ResponseError) {
          console.error("HTTP status code: " + e.status);

        }
        problemConnecting();

      }

    }


    requestModel();




  };

  useEffect(() => {

    obtainData();
  }, []);


  const updateCurLoc = (locationName) => {
    setCurLoc(locationMap.get(locationName));
  }

  const markerClicked = (location) => {
    updateCurLoc(location);
  }

  const _onLocationSelect = (event) => {

    updateCurLoc(event.target.value);

  }

  const _onVarSelect = (event) => {

    setCurVar(event.target.value);

  }

  const handleStartChange = (selectedDate) => {
    setFromDate(formatDate(selectedDate, FORMAT));

  }

  const handleEndChange = (selectedDate) => {

    setToDate(formatDate(selectedDate, FORMAT));

  }

  const spaceToUnderscore = (s) => {
    return s.replace(/ /g, '_')
  }


  const handleCSV = async () => {

    const response = await fetch(expandLink(curLoc, curVar, fromDate, toDate));

    if (response.ok) {
      const blob = await response.blob();
      const a = document.createElement('a');
      a.href = window.URL.createObjectURL(blob);
      a.download = spaceToUnderscore(curLoc.name) + '_' + spaceToUnderscore(curVar) + '_'
        + fromDate + "_" + toDate + '.csv';
      a.click();
    } else if (response.status === HttpStatus.GATEWAY_TIMEOUT) {
      problemConnecting();
    }
  }

  return (

    <div id="mapdiv">

      <div id="logos">

        <center>
          {props.logoLeft}
          {props.logoRight}
        </center>

      </div>

      {props.createMap(marker, curVar, locationMap, markerClicked, fromDate, toDate, curLoc)}

      <div id="selectdiv" >

        <center>

          <div>
            {curLoc && <div id='variDD' >

              {varOpt && <select onChange={_onVarSelect}>

                {varOpt.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

              </select>}

            </div>}

            {curLoc && <div className='marginItem' >

              <select onChange={_onLocationSelect} value={curLoc.name}>
                {!option && <option key="Location" value="Location" >Location</option>}
                {option && option.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

              </select>

            </div>}


            <div className='pLeft' >{
              createSelector("From date", fromDate, handleStartChange)
            }
            </div>

            <div className='pLeft' >
              {
                createSelector("To date", toDate, handleEndChange)
              }
            </div>

            {curLoc && curVar && <div className='pLeft'> <button onClick={handleCSV}>CSV</button> </div>}
          </div>

          <ARLocationComponent createLocation={createLocationFactory(obtainData, collectionModel)}
            createRemove={curLoc ? createRemoveFactory(obtainData, curLoc._links[SELF].href, curLoc.name) : null} />

        </center>

      </div>



    </div>
  );

}

Unison.propTypes = {

  /** A function for creating the map. */
  createMap: PropTypes.func.isRequired,

  /** A logo displayed at the bottom of the screen. It will be displayed to the left
   * if the logoRight prop is defined.
   */
  logoLeft: PropTypes.any,

  /** A logo displayed at the bottom of the screen. It will be displayed to the right
   * if the logoLeft prop is defined.
   */
  logoRight: PropTypes.any,

}

export default Unison;
