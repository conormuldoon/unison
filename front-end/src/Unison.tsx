
import React, { useState, useEffect } from 'react';
import 'react-day-picker/lib/style.css';
import { formatDate } from 'react-day-picker/moment';
import './App.css';
import ARLocationComponent from './ARLocationComponent';
import { FORMAT, SELF } from './Constant';
import DateSelector from './DateSelector';
import { today, tomorrow, problemConnecting, varMapping } from './Util';
import HttpStatus from 'http-status-codes';

import { MapMarker } from './LeafletMap';

import { createLocationFactory } from './LocationForm';

import { createRemoveFactory } from './RemoveComponent';
import { createPopupFactory } from './ChartPopup';
import parser from 'uri-template';


interface UnisonProps {

  createMap: (marker: MapMarker[] | null,
    markerCallback: (locationName: string) => void,
    popupFactory?: (closePopup: () => void) => React.ReactNode) => JSX.Element | undefined;

  logoLeft?: React.ReactNode;
  logoRight?: React.ReactNode;
}


type ModelLink = {
  "self": string,
  "contains": string

}



class ResponseError extends Error {

  private status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }

  logStatus() {

    console.error("HTTP status code: " + status);

  }
}

function checkResponse(ok: boolean, endpoint: string, status: number) {
  if (!ok)
    throw new ResponseError('Bad response for ' + endpoint, status);
}

const OMIT = 'omit';

function jsonHeader(accept: 'geo' | 'hal'): RequestInit {
  return {
    credentials: OMIT,
    headers: new Headers({
      'Accept': 'application/' + accept + '+json'
    })
  }
}

const halGetHeader: RequestInit = jsonHeader('hal');

type Link = {
  [key: string]: {
    href: string
  };
}

function createLinkMap(listItem: Link) {
  const itemMap = new Map<string, string>();
  for (const rel in listItem) {
    itemMap.set(rel, listItem[rel].href);

  }
  return itemMap;
}

const spaceToUnderscore = (s: string) => {
  return s.replace(/ /g, '_')
}

function optionName(s: string) {
  let optName = '';
  let sw = 0;
  for (let i = 1; i < s.length; i++) {
    const c = s.charAt(i);
    if (c >= 'A' && c <= 'Z') {
      optName += s.substring(sw, i) + ' ';
      sw = i;
    }
  }
  optName += s.substring(sw);
  return optName;
}

function variableOptions(vo: string[], link: Link) {
  for (const rel in link) {
    if (SELF === rel) {
      continue;
    }

    const s = rel.charAt(0).toUpperCase() + rel.substring(1);
    const optName = optionName(s);
    vo.push(optName);
  }
}

type LinkMap = Map<string, string>;


/**
 * Application component for Unison. Once mounted, it connects to the back-end to receive a list of the locations being tracked.
 * 
 * @component
 * 
 */
function Unison({ createMap, logoLeft, logoRight }: UnisonProps): JSX.Element {


  const [fromDate, setFromDate] = useState(today());
  const [toDate, setToDate] = useState(tomorrow());
  const [option, setOption] = useState<string[] | null>(null);
  const [marker, setMarker] = useState<MapMarker[] | null>(null);
  const [curLoc, setCurLoc] = useState<string | null>(null);
  const [curVar, setCurVar] = useState<string | null>(null);
  const [varOpt, setVarOpt] = useState<string[] | null>(null);
  const [locationMap, setLocationMap] = useState<Map<string, LinkMap> | null>(null);
  const [modelLink, setModelLink] = useState<ModelLink | null>(null);

  async function requestFeatureCollection(uri: string) {
    const response = await fetch(uri, jsonHeader('geo'));


    checkResponse(response.ok, uri, response.status);

    const fc = await response.json();

    const locationArray = fc.features;
    const n = locationArray.length;

    const newOption = [];
    const newMarker: MapMarker[] = [];

    for (let i = 0; i < n; i++) {
      newOption.push(locationArray[i].properties.name);
      const pos: [number, number] = [locationArray[i].geometry.coordinates[1], locationArray[i].geometry.coordinates[0]];

      const properties = locationArray[i].properties;
      newMarker.push({ name: properties.name, position: pos });

    }

    if (n > 0) {

      setOption(newOption);
      setMarker(newMarker);


    } else {
      setOption(null);
      setMarker(null);
    }


  }

  function clearCurrent() {

    setCurLoc(null);
    setCurVar(null);
    setVarOpt(null);
    setLocationMap(null);

  }

  function processModelList(list: [{ name: string, _links: Link }]) {

    const n = list.length;
    const map = new Map();

    for (let i = 0; i < n; i++) {

      map.set(list[i].name, createLinkMap(list[i]._links));
    }

    setLocationMap(map);

    const vo: string[] = [];

    if (n > 0) {
      variableOptions(vo, list[0]._links);

    }

    if (vo.length > 0) {
      setCurLoc(list[0].name);
      setCurVar(vo[0]);
      setVarOpt(vo);

    } else {

      clearCurrent();
      alert("No weather data is associated with the collection.")
    }
  }

  async function requestCollectionHAL(uri: string) {

    const response = await fetch(uri, halGetHeader);


    checkResponse(response.ok, uri, response.status);

    const model = await response.json();


    setModelLink({ self: model._links.self.href, contains: model._links.contains.href });


    if (model._embedded) {
      processModelList(model._embedded.locationModelList);
    } else {
      clearCurrent();
    }


  }

  const obtainData = () => {


    async function requestModel() {
      try {

        const root = process.env.PUBLIC_URL;

        const response = await fetch(root, jsonHeader('hal'));

        checkResponse(response.ok, root, response.status);
        const model = await response.json();
        const uri = model._links.locationCollection.href;
        await requestCollectionHAL(uri);
        await requestFeatureCollection(uri);

      } catch (e) {
        if (e instanceof ResponseError) {
          e.logStatus();

        } else {
          console.log(e);
        }
        problemConnecting();


      }

    }

    requestModel();




  };

  useEffect(obtainData, []);

  const markerClicked = (location: string) => {
    setCurLoc(location);

  }

  const _onLocationSelect = (event: React.ChangeEvent<HTMLSelectElement>) => {

    setCurLoc(event.target.value);

  }

  const _onVarSelect = (event: React.ChangeEvent<HTMLSelectElement>) => {

    setCurVar(event.target.value);

  }

  const handleStartChange = (selectedDate: Date) => {
    setFromDate(formatDate(selectedDate, FORMAT));

  }

  const handleEndChange = (selectedDate: Date) => {

    setToDate(formatDate(selectedDate, FORMAT));

  }



  function expandLink() {

    if (locationMap && curLoc && curVar) {
      const linkMap = locationMap.get(curLoc);
      if (linkMap) {
        const href = linkMap.get(varMapping(curVar));
        const template = parser.parse(href);
        return template.expand({ name: curLoc, fromDate: fromDate, toDate: toDate });
      }
    }


  }

  const handleCSV = async () => {


    const response = await fetch(expandLink(), {
      credentials: OMIT
    });

    if (response.ok && curVar && curLoc) {
      const blob = await response.blob();
      const a = document.createElement('a');
      a.href = window.URL.createObjectURL(blob);
      a.download = spaceToUnderscore(curLoc) + '_' + spaceToUnderscore(curVar) + '_'
        + fromDate + "_" + toDate + '.csv';
      a.click();
    } else if (response.status === HttpStatus.GATEWAY_TIMEOUT) {
      problemConnecting();
    }
  }


  function createSelector(label: string, dateValue: Date, handleDayChange: (day: Date) => void) {
    return <DateSelector label={label} dateValue={dateValue} handleDayChange={handleDayChange} />;
  }

  let popupFactory;
  if (curLoc && curVar) {
    const uri = expandLink();
    popupFactory = createPopupFactory(uri, curVar, curLoc);
  }

  let arc = null;
  if (modelLink) {
    let createRemove = null;


    if (curLoc && locationMap) {

      const linkMap = locationMap.get(curLoc);
      if (linkMap) {
        createRemove = createRemoveFactory(obtainData, curLoc, linkMap.get(SELF));
      }
    }
    arc = <ARLocationComponent createLocation={createLocationFactory(obtainData, modelLink.self, modelLink.contains)}
      createRemove={createRemove} />
  }

  return (

    <div id="mapdiv">

      <div id="logos" style={{ textAlign: "center" }} >

        {logoLeft}
        {logoRight}


      </div>

      {createMap(marker, markerClicked, popupFactory)}

      <div id="selectdiv" style={{ textAlign: "center" }}>

        <div>
          {curLoc && <div id='variDD' >

            {varOpt && <select onChange={_onVarSelect}>

              {varOpt.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

            </select>}

          </div>}

          {curLoc && <div className='marginItem' >

            <select onChange={_onLocationSelect} value={curLoc}>
              {!option && <option key="Location" value="Location" >Location</option>}
              {option && option.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

            </select>

          </div>}


          <div className='pLeft' >{
            createSelector("From date", fromDate, handleStartChange)
          }
          </div>

          <div className='pLeft' >
            {
              createSelector("To date", toDate, handleEndChange)
            }
          </div>


          {curLoc && curVar && <div className='pLeft'> <button onClick={handleCSV}>CSV</button> </div>}
        </div>

        {arc}


      </div>

    </div>
  );

}


export default Unison;
