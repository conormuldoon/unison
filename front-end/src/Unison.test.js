import "@testing-library/jest-dom/extend-expect";
import fetchMock from 'fetch-mock';
import React from 'react';
import ReactDOM from 'react-dom';
import { fireEvent, render, waitForElement } from "react-testing-library";
import Unison from './Unison';

const features = { "features": [{ "geometry": { "type": "Point", "coordinates": [-6.223682, 53.308441] }, "properties": { "name": "UCD" } }] };

it('renders without crashing', async () => {


    fetchMock.get('end:/location', features);

    const div = document.createElement('div');

    ReactDOM.render(<Unison mapCentre={[59.922326, 10.751560]} />, div);
    ReactDOM.unmountComponentAtNode(div);

    fetchMock.restore();
});

it('mathes Unison snapshot', () => {


    fetchMock.get('end:/location', features);
    const mockDateNow = jest.fn(() => 1571875200000);
    const dn = global.Date.now;
    global.Date.now = mockDateNow;

    const { container } = render(<Unison mapCentre={[59.922326, 10.751560]} />);

    expect(container).toMatchSnapshot();
    global.Date.now = dn;
    fetchMock.restore();

});

it('displays popup when marker clicked', async () => {
    fetchMock.get('end:/location', features);


    fetchMock.get('end:/precipitation?location=UCD&fromDate=1/9/2019&toDate=23/10/2019', [{
        "date": "2019-04-01T23:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0.1
        }
    },
    {
        "date": "2019-04-02T00:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T01:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T02:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T03:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T04:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T05:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T06:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T07:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T08:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T09:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T10:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    }]);

    const { getAllByAltText, getByText /*, debug*/ } = render(<Unison mapCentre={[59.922326, 10.751560]} />);

    const marker = await waitForElement(() => getAllByAltText('')[1]);

    fireEvent.click(marker);
    const text = await waitForElement(() => getByText('UCD'));

    expect(text).toBeDefined();
    //debug();
    fetchMock.restore();
});

