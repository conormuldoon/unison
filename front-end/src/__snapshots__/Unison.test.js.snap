// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathes Unison snapshot 1`] = `
<div>
  <div
    id="mapdiv"
  >
    <div
      id="logos"
    >
      <center>
        <img
          alt=""
          id="logoitem"
        />
        <img
          alt=""
          id="logoitem"
        />
      </center>
    </div>
    <div
      class="leaflet-container leaflet-grab leaflet-touch-drag"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://b.tile.osm.org/12/2170/1191.png"
                style="width: 256px; height: 256px; left: -84px; top: -62px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        />
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              +
            </a>
            <a
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              −
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | © 
            <a
              href="https://osm.org/copyright"
            >
              OpenStreetMap
            </a>
             contributors
          </div>
        </div>
      </div>
      <div
        id="marginclickdiv"
      />
    </div>
    <div
      id="selectdiv"
    >
      <center>
        <div>
          <div
            id="variDD"
          >
            <select
              disabled=""
            >
              <option
                value="Precipitation"
              >
                Precipitation
              </option>
              <option
                value="Humidity"
              >
                Humidity
              </option>
              <option
                value="Wind Direction"
              >
                Wind Direction
              </option>
              <option
                value="Wind Speed"
              >
                Wind Speed
              </option>
              <option
                value="Cloudiness"
              >
                Cloudiness
              </option>
              <option
                value="Cloud Level"
              >
                Cloud Level
              </option>
              <option
                value="Dew Point"
              >
                Dew Point
              </option>
              <option
                value="Pressure"
              >
                Pressure
              </option>
              <option
                value="Temperature"
              >
                Temperature
              </option>
            </select>
          </div>
          <div
            class="marginItem"
          >
            <select
              disabled=""
            >
              <option
                selected=""
                value="Location"
              >
                Location
              </option>
            </select>
          </div>
          <div
            class="pLeft"
          >
            <div>
              <p>
                From date
              </p>
              <div
                class="DayPickerInput"
              >
                <input
                  placeholder="YYYY-M-D"
                  value="24/10/2019"
                />
              </div>
            </div>
          </div>
          <div
            class="pLeft"
          >
            <div>
              <p>
                To date
              </p>
              <div
                class="DayPickerInput"
              >
                <input
                  placeholder="YYYY-M-D"
                  value="25/10/2019"
                />
              </div>
            </div>
          </div>
          <a
            class="pLeft"
            href="/apiacc/csvPrecipitation_1571875200000.csv?location=undefined&fromDate=24/10/2019&toDate=25/10/2019"
          >
            <button
              disabled=""
            >
              CSV
            </button>
          </a>
        </div>
        <div>
          <div
            class="pLeft"
          >
            <div
              id="pLeft"
            >
              <button
                data-testid="lf-button"
              >
                Add Location
              </button>
            </div>
          </div>
        </div>
      </center>
    </div>
  </div>
</div>
`;
